scraper/cazy_webscraper.py

```diff
@@ -539,13 +517,42 @@ def parse_family_pages(family_url, family_name, cazy_home, logger):
                 f"No protein records for CAZy family {family_name} will be retried."
             )
         )
-        return []
+        return family
```

Comment:
Same query/possible improvement as above.

Reply:
I was having a think about this.

I'm currently adding in so that a user can define how many times to retry connecting to CAZy when scraping the home pages, class pages or family "all" page.

This is separate to the retries in the decorator that wraps the function that makes the actual call to CAZy. The "--retries" option allows the user to set the number of times to try the series of 10 (although this may be reduced, potentially to 3) attempted connections (performed within the decorator) to CAZy.

I'm torn between:
A.
If an attempt to scrape the current target page fails, immediately retry the scrape attempt for the current target webpage
B.
Add the URL to the end of URL list that is currently being iterated through, so there is a break between attempts to scrape the target page. For example, this would add the family URL to the family URL list that is iterated through in get_cazy_data(). The "--retries" option would determine how many times a failed scrape of the family/class will result in the family/class being added to the back of the list of times to scrape.

Option A is the quickest and easiest to code. Option B will require a method to associate the target page URL (or other identifier) with the number of times attempts to scrape the page have been made.

Option B wouldn't necessarily be too much more work, the main issue is selecting a suitable method for storing the data (associating the item to scrape with the number of attempted scrapes). Simplest could a nested list with the order of elements as:
 [<URL or page name>, <number of times attempts so scrape the page have been performed>]
Or a dictionary keyed but the URL/page name and values by the number of time attempts to scrape the page has been performed.
Or a dataclass is created for the CAZy classes, including the Name, URL and 'attempted_scrapes', and add the number of attempts added to the Family class. For the Families I'm leaning towards the last option, I'm still debating the option for handling scraping of the class pages

When scraping a CAZy family it is possible that the scraping of some protein table pages is successful and for others fails. To save having to scrape protein table pages that have been successfully scraped I will store the URL of the specific protein table pages that were not successfully scrapped (meaning a connection to CAZy was not made) in the Family class object, and only the retry the connection to CAZy only for those specified URLs.

This would lead to option B conflicting with the current output method of pandas df, because if the option to split data by families is enabled as soon as a Family is scraped the data is written out to a data frame and that is written out to a .CSV file. Therefore, adding the family URL to the end of the list of family URLs would result in multiple output dataframes for each family. **But** seeing as we're moving to an SQL database in the long run I think this is fine, because I'm thinking the behaviour of the scraper will be that all protein data will be added to a single SQL database and if the user wants class or family specific data they can query it to retrieve the data they wanted. Or to make it easier for the user the data can be split for them (as specified by the --data_split option, but this will be done by querying the database for the user once all the scraping is done.

I'm also thinking (actually very likely to go ahead) of adding the writing out a separate log file (or other output file type) listing the classes/families for which the scrape failed. The file would contain something along the lines "The following requested items were not scrapped from CAZy" and then list the pages for which the scrape failed. This is because, atm the failure to scrape a class/family is logged and written out to the terminal, but this could be difficult for user to retrieve if they are performing a very large scrape.